ros2/test/pluginlib/ament_setup_library_test.cc

```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 Wouter van der Stoel
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "console_bridge/console.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "pluginlib/class_loader.hpp"
+#include "rcutils/logging.h"
+
+#include "ros2/test/pluginlib/ament_setup.hpp"
+#include "ros2/test/pluginlib/regular_polygon.h"
+
+using ::testing::DoubleNear;
+using ::testing::Eq;
+
+TEST(TestAmentSetup, PluginLoadingDoesNotWorkUntilAmentPrefixPathIsSet) {
+  rcutils_logging_set_default_logger_level(RCUTILS_LOG_SEVERITY_DEBUG);
+  console_bridge::setLogLevel(console_bridge::CONSOLE_BRIDGE_LOG_DEBUG);
+
+  EXPECT_ANY_THROW(
+      pluginlib::ClassLoader<polygon_base::RegularPolygon> poly_loader(
+          "polygon_base", "polygon_base::RegularPolygon"));
+}
+
+TEST(TestAmentSetup, SetupAmentPrefixPath) {
+  ::ros2_test_pluginlib_ament_setup::setup_ament_prefix_path();
```

Comment:
Is it guaranteed that this init is definitely going to happen before the consumption of the prefix path? (which also happens during static var init phase)

Reply:
You are correct that this function needs to be called very early. Are you aware of any initialization that happens prior to a ROS2 function being called from user code? In my use case of reading rosbags from an otherwise "pure" C++ library explicitly calling this function prior to calling a rosbags API function is sufficient.

Reply:
I searched a bit, but couldn't find anything solid on this topic. There is no dependency in static variables in this case that could cause [Static Initialization Order Fiasco](https://en.cppreference.com/w/cpp/language/siof). Here we have to rely on ordering of init of static vars. TBH, I am surprised this works.

@ahans WDYT?

Reply:
I think this is fine. `pluginlib` and `class_loader` maintain some global state, but that is only for already loaded libraries. If you try to load a plugin explicitly as in the present test or indirectly by requesting type support for some message type, eventually `ament_index_cpp` will be queried which does not cache anything, but always looks at the environment variable.

That being said, the shell script wrapper solution is certainly the cleaner one. It makes sure the environment variable is set from the start of the executable and doesn't change during runtime. This way it doesn't depend on a ROS implementation detail. If for some reason ament maintainers figure it's a good idea to cache some results and not read the environment variable each time (maybe tokenization of `AMENT_PREFIX_PATH` is considered too expensive...), the wrapper approach would still work, while the present one (changing the environment variable from within the executable) would break. That's not to say I'm against it. After all, before the shell script wrapper we had very similar functionality only available for Python targets and I argued at some point it would make sense to also have that for C++ targets. But one must be aware that there are certain caveats.

Reply:
OK, let's then proceed with the high-level idea of this PR. It should be well documented that the call to `setup_ament_prefix_path ` should happen "early enough" such that plugin registration works.